components/director/internal/domain/onetimetoken/service.go

```diff
@@ -236,13 +239,25 @@ func (s *service) getTokenFromAdapter(ctx context.Context, adapterURL string, ap
 	}
 
 	extTenant := tntCtx.ExternalID
+	var tnt *model.BusinessTenantMapping
 	if len(extTenant) == 0 {
-		extTenant, err = s.extTenantsSvc.GetExternalTenant(ctx, tntCtx.InternalID)
-		if err != nil {
-			return nil, errors.Wrapf(err, "while getting external tenant for internal tenant [%s]", tntCtx.InternalID)
+		if tnt, err = s.extTenantsSvc.GetTenantByID(ctx, tntCtx.InternalID); err != nil {
+			return nil, errors.Wrapf(err, "while getting tenant with internal ID %q", tntCtx.InternalID)
+		}
+	} else {
+		if tnt, err = s.extTenantsSvc.GetTenantByExternalID(ctx, tntCtx.ExternalID); err != nil {
+			return nil, errors.Wrapf(err, "while getting tenant with external ID %q", tntCtx.ExternalID)
 		}
 	}
 
+	if tnt.Type == tenantpkg.Subaccount {
+		if extTenant, err = s.extTenantsSvc.GetExternalTenant(ctx, tnt.Parent); err != nil {
+			return nil, errors.Wrapf(err, "while getting parent external tenant for internal tenant %q", tnt.Parent)
+		}
+	} else {
+		extTenant = tnt.ExternalTenant
```

Comment:
Maybe set it by default to `tnt.ExternalTenant` above the check for subaccount? I think this might be a bit easier for readability.